include/revng/TupleTree/Tracking.h

```diff
@@ -0,0 +1,225 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <cstdint>
+#include <tuple>
+
+#include "revng/ADT/KeyedObjectContainer.h"
+#include "revng/ADT/TrackingContainer.h"
+#include "revng/ADT/UpcastablePointer.h"
+#include "revng/Support/YAMLTraits.h"
+#include "revng/TupleTree/TupleLikeTraits.h"
+#include "revng/TupleTree/TupleTreeCompatible.h"
+#include "revng/TupleTree/TupleTreePath.h"
+#include "revng/TupleTree/Visits.h"
+
+struct ReadFields {
+  std::set<TupleTreePath> Read;
+  std::set<TupleTreePath> ExactVectors;
+};
+
+namespace revng::detail {
+
+template<typename M>
+struct TrackingImpl {
+
+  ReadFields collect(const M &LHS) {
+    TupleTreePath Stack;
+    ReadFields Info;
+    collectImpl(LHS, Stack, Info);
+    return Info;
+  }
+
+  void clear(const M &LHS) { clearImpl(LHS); }
+
+  void push(const M &LHS) { pushImpl(LHS); }
+
+  void pop(const M &LHS) { popImpl(LHS); }
+
+private:
+  template<size_t I = 0, typename T>
+  void collectTuple(const T &LHS, TupleTreePath &Stack, ReadFields &Info) {
+    if constexpr (I < std::tuple_size_v<T>) {
+
+      Stack.push_back(size_t(I));
+      if (LHS.template getTracker<I>().isSet())
+        Info.Read.insert(Stack);
+      collectImpl(LHS.template untrackedGet<I>(), Stack, Info);
+      Stack.pop_back();
+
+      // Recur
+      collectTuple<I + 1>(LHS, Stack, Info);
+    }
+  }
+
+  template<StrictSpecializationOf<UpcastablePointer> T>
+  void collectImpl(const T &LHS, TupleTreePath &Stack, ReadFields &Info) {
+    LHS.upcast(
+      [&](auto &LHSUpcasted) { collectImpl(LHSUpcasted, Stack, Info); });
+  }
+
+  template<TupleSizeCompatible T>
+  void collectImpl(const T &LHS, TupleTreePath &Stack, ReadFields &Info) {
+    collectTuple(LHS, Stack, Info);
+  }
+
+  template<revng::detail::SetOrKOC T>
+  void collectImpl(const T &LHS, TupleTreePath &Stack, ReadFields &Info) {
+    typename T::TrackingResult TrackingResult = LHS.getTrackingResult();
+    if (TrackingResult.Exact)
+      Info.ExactVectors.insert(Stack);
+
+    for (auto &Key : TrackingResult.InspectedKeys) {
+      Stack.push_back(Key);
+      Info.Read.insert(Stack);
+      Stack.pop_back();
+    }
+    for (auto &LHSElement : LHS.Content) {
+      using value_type = typename T::value_type;
+
+      if constexpr (TupleSizeCompatible<value_type>)
+        Stack.push_back(LHSElement.untrackedKey());
+      else
+        Stack.push_back(KeyedObjectTraits<value_type>::key(LHSElement));
+
+      collectImpl(LHSElement, Stack, Info);
+      Stack.pop_back();
+    }
+  }
+
+  template<NotTupleTreeCompatible T>
+  void collectImpl(const T &LHS, TupleTreePath &Stack, ReadFields &Info) {}
+
+private:
+  template<size_t I = 0, typename T>
+  void clearTuple(const T &LHS) {
+    if constexpr (I < std::tuple_size_v<T>) {
+
+      clearImpl(LHS.template untrackedGet<I>());
+      LHS.template getTracker<I>().clear();
+
+      // Recur
+      clearTuple<I + 1>(LHS);
+    }
+  }
```

Comment:
Collapse implementations of `clear`, `push` and `pop` in a single one taking a lambda or something.